csharp/src/Ice/Reference.cs

```diff
@@ -460,16 +532,30 @@ public virtual bool Equals(Reference? other)
                                                 nameof(invocationTimeout));
                 }
 
-                if (reference == this)
-                {
-                    reference = Clone();
-                }
+                reference = Clone(reference);
                 reference._invocationTimeout = invocationTimeoutValue;
             }
             return reference;
         }
 
-        public Reference Clone() => (Reference)MemberwiseClone();
+        public Reference Clone()
+        {
+            var clone = (Reference)MemberwiseClone();
+            clone._requestHandler = null;
+            clone._requestHandlerMutex = null;
+            return clone;
+        }
+
+        // Creates a new clone of this unless `clone` is already a real clone (meaning "not this")
+        // Returns the clone.
+        private protected T Clone<T>(T clone) where T : Reference
```

Comment:
A more descriptive name would be less confusing, perhaps CloneIfSame or CloneIfReferenceEqual?

Reply:
or maybe just
```
reference = reference == this ? Clone() : reference;
```

I found the previous code clear that this clone overload.

Reply:
The previous code was actually incorrect: `==` is a deep value-like comparison, while the correct comparison here is `ReferenceEquals` (we're checking if reference was already cloned).

The old code was also quite repetitive, with lots of:
```
if (reference == this) // should be ReferenceEquals
{
    reference = (RoutableReference)Clone();
}
```
Surely a helper method should take care of that.